test/helm/chart-test.sh

```diff
@@ -0,0 +1,229 @@
+#!/bin/bash
+
+# Script to test helm charts for the following:
+## YAML validity
+## chart deployability
+## chart version increment when chart is changed
+## chart installation on Kubernetes
+
+# Prerequisites:
+## Docker
+
+set -euo pipefail
+
+# KIND / Kubernetes
+KIND_VERSION="v0.8.1"
+readonly K8s_1_18="v1.18.2"
+KIND_IMAGE="$K8s_1_18"
+readonly CLUSTER_NAME="kind-ct"
+readonly REPO_PATH="$( cd "$(dirname "$0")"; cd ../../ ; pwd -P )"
+readonly CLUSTER_CONFIG="$REPO_PATH/test/helm/kind-config.yaml"
+readonly KUBECONFIG_TMP_DIR="$REPO_PATH/build/tmp-$CLUSTER_NAME"
+readonly KUBECONFIG_TMP_PATH="$KUBECONFIG_TMP_DIR/kubeconfig"
+readonly KIND_EXEC_ARGS="--context kind-$CLUSTER_NAME --kubeconfig $KUBECONFIG_TMP_PATH"
+
+# Helm/chart-testing
+CT_TAG="v3.0.0-rc.1"
+readonly CT_CONFIG="test/helm/ct.yaml"
+
+# chart-testing container
+readonly CT_CONTAINER_NAME="ct"
+readonly CT_EXEC="docker exec --interactive $CT_CONTAINER_NAME"
+
+readonly HELP=$(cat << 'EOM'
+Test Helm charts by linting and installing helm charts in a provisioned environment. Only changed charts are tested.
+Provisioned environment includes a Kind Kubernetes cluster and Docker container with helm/chart-testing image.
+
+Usage:
+  chart-test [options]
+
+Examples:
+  chart-test -l                       Run the test for linting only for default Kubernetes version
+  chart-test -k v1.17.5               Run the test for linting and installation for Kubernetes v1.18.2
+  chart-test -r -c v3.0.0-rc.1        Run the test for linting and installation for default Kubernetes version, and reuse previously provisioned test environment
+
+Options:
+  -k     Kubernetes version / kindest node image tag to use for the test
+  -c     chart-testing image tag to use for the test
+  -l     test charts for linting only (helm lint, version checking, YAML validation, maintainer validation)
+  -p     preserve the provisioned environment after test runs
+  -r     reuse kind cluster and docker chart-testing container previously provisioned by this tool
+  -d     debug, enables set -x, printing primary commands before executing
+  -h     help message
+EOM
+)
+
+LINT_ONLY=false
+DEBUG=false
+PRESERVE=false
+REUSE_ENV=false
+
+RESOURCES_CREATED=[]
+
+RED=$(tput setaf 1)
+GREEN=$(tput setaf 2)
+MAGENTA=$(tput setaf 5)
+RESET_FMT=$(tput sgr 0)
+BOLD=$(tput bold)
+
+setup_ct_container() {
+    c_echo "Provisioning and running chart-testing container named $CT_CONTAINER_NAME..."
+    docker run --rm --interactive --detach --network host --name $CT_CONTAINER_NAME \
+        --volume "$REPO_PATH/$CT_CONFIG:/etc/ct/ct.yaml" \
+        --volume "$REPO_PATH:/workdir" \
+        --workdir /workdir \
+        "quay.io/helmpack/chart-testing:$CT_TAG"
+    echo
+}
+
+install_kind() {
+    c_echo "Installing kind..."
+    curl -Lo ./kind https://kind.sigs.k8s.io/dl/$KIND_VERSION/kind-$(uname)-amd64
+    chmod +x ./kind
+    mv ./kind /usr/local/bin/kind
```

Comment:
I think it's best to `mv` the kind binary to the tmp dir build directory rather than `/usr/local/bin/`. Since we're installing kind all the time now, it might overwrite user's installed version and depending on the PATH priority another version of Kind could still be used.  You can setup your own PATH priority for the duration of this script to first look in the tmp build dir.

Reply:
makes sense. 👍